File: src/Daemon/Checker.cs
using System;
using System.Data;
using System.Threading;
using System.Collections.Generic;

using ServerCheck.Network;

namespace ServerCheck.Daemon
{
    sealed class Checker
    {
        private static readonly Checker _instance = new Checker();

        private Checker()
        {
        } 

        public static Checker GetInstance()
        {
            return _instance;
        }

        private bool CheckConnection(string address_type, string address)
        {
            return address_type switch
            {
                "ip" => Request.PingRequest(address),
                "url" => Request.HttpRequest(address, "GET"),
                _ => false
            };
        }

        public void Run(int checkTime)
        {
            HashSet<Tuple<string, string>> notified = new HashSet<Tuple<string, string>>();
            Database.MySql db = new Database.MySql();

            while (true)
            {
                DataTable data = db.GetRows(
                    "s.address, s.address_type, a.contact, a.contact_type",
                    "servers s JOIN observes o ON s.address = o.address JOIN accounts a ON a.contact = o.contact"
                );

                foreach (DataRow row in data.Rows)
                {
                    string serverAddress = row["address"].ToString();
                    string serverAddressType = row["address_type"].ToString();
                    string receiverContact = row["contact"].ToString();
                    string receiverContactType = row["contact_type"].ToString();

                    string checkTimestamp = DateTime.UtcNow.ToString("s") + " UTC";
                    bool isConnected = this.CheckConnection(serverAddressType, serverAddress);

                    Tuple<string, string> serverObserver = new Tuple<string, string>(serverAddress, receiverContact);
                    bool wasUnhealthy = notified.Contains(serverObserver);
                                     
                    if ((isConnected && wasUnhealthy) || (!isConnected && !wasUnhealthy))
                    {
                        AlertMessage newAlert = new AlertMessage(isConnected, serverAddress, checkTimestamp);
                        newAlert.Notify(receiverContact, receiverContactType);
                        
                        if (isConnected && wasUnhealthy) notified.Remove(serverObserver);
                        else if (!isConnected && !wasUnhealthy) notified.Add(serverObserver);
                    }
                }

                Thread.Sleep(checkTime * 1000);
            }
        }

    }
}

File: app/src/main/java/com/antonioscardace/app/Daemon/Checker.java
package com.antonioscardace.app.Daemon;

import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;
import java.util.TimeZone;

import javafx.util.Pair;

import javax.mail.MessagingException;

import com.antonioscardace.app.Config;
import com.antonioscardace.app.Database.MySql;
import com.antonioscardace.app.Request.IRequest;
import com.antonioscardace.app.Request.RequestCreator;

public final class Checker {
    private final Integer CHECK_TIME;
    private Set<Pair<String, String>> notified;

    public Checker() {
        this.CHECK_TIME = Config.getInstance().getInt("checkTime");
        this.notified = new HashSet<>();
    }

    private boolean checkServer(String address, String addressType) throws IOException {
        IRequest req = RequestCreator.getRequest(addressType);
        return req.request(address);   
    }

    private String getUtcTimestamp() { 
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"); 
		dateFormat.setTimeZone(TimeZone.getTimeZone("UTC")); 
		return dateFormat.format(new Date()) + " UTC";
	}

    private void updateNotifics(boolean isConnected, boolean wasUnhealthy, Pair<String, String> observer) {
        if (isConnected && wasUnhealthy) this.notified.remove(observer);
        else if (!isConnected && !wasUnhealthy) this.notified.add(observer);
    }

    private void scanServer(String address, String addressType, String contact, String contactType) throws IOException, MessagingException {
        String timestamp = getUtcTimestamp();
        boolean isConnected = this.checkServer(address, addressType);

        Pair<String, String> observer = new Pair<String,String>(address, contact);
        boolean wasUnhealthy = this.notified.contains(observer);

        if ((isConnected && wasUnhealthy) || (!isConnected && !wasUnhealthy)) {
            AlertMessage newAlert = new AlertMessage(isConnected, address, timestamp);
            newAlert.notify(contact, contactType);
            this.updateNotifics(isConnected, wasUnhealthy, observer);
        }
    }

    public void run() throws ClassNotFoundException, SQLException, InterruptedException, IOException, MessagingException {
        Config config = Config.getInstance();
        MySql db = new MySql(
            config.get("server"),
            config.getInt("port"),
            config.get("userid"),
            config.get("password"),
            config.get("dbName")
        );

        while (true) {
            ResultSet data = db.getRows(
                "servers s JOIN observes o ON s.address = o.address JOIN accounts a ON a.contact = o.contact",
                "s.address, s.address_type, a.contact, a.contact_type"
            );
            
            while (data.next()) 
                this.scanServer(
                    data.getString("address"),
                    data.getString("address_type"),
                    data.getString("contact"),
                    data.getString("contact_type")
                );
                
            Thread.sleep(this.CHECK_TIME * 1000);
        }
    }
}